### PR TITLE
Machine ID: Add Workload Identity Service

### PR DESCRIPTION
**Pull request by strideynet**
Part of https://github.com/gravitational/teleport/issues/36205

As per RFD https://github.com/gravitational/teleport.e/blob/master/rfd/0018e-machine-id-workload-identity-mvp.md

This PR is an element broken off of https://github.com/gravitational/teleport/pull/37046 if you desire additional context. There are other PRs in this set, so this may appear "useless" on its own.

Depends on https://github.com/gravitational/teleport/pull/38174
Depends on https://github.com/gravitational/teleport/pull/38172

### CODE REVIEW

**Comment by strideynet** [comment]
Draft until dependent PRs merge and we can rebase this on master.

**Comment by strideynet** [comment]
@capnspacehook, @jakule, @timothyb89 - are you able to take a look at this this week ?

File: lib/auth/machineid/machineidv1/experiment/experiment.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package experiment
+
+import (
+	"os"
+	"sync"
+)
+
+var mu sync.Mutex
```
Comment (jakule) [inline lib/auth/machineid/machineidv1/experiment/experiment.go:26]: If you only need bool, you can use atomic for that.

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+			user: unauthorizedUser.GetName(),
+			req: &machineidv1pb.SignX509SVIDsRequest{
+				Svids: []*machineidv1pb.SVIDRequest{
+					{
+						SpiffeIdPath: "/alpha/foo",
+						PublicKey:    pubBytes,
+					},
+				},
+			},
+			requireError: func(t require.TestingT, err error, i ...interface{}) {
```
Comment (jakule) [inline lib/auth/machineid/machineidv1/workload_identity_service_test.go:208]: nit:
```suggestion
			requireError: func(t require.TestingT, err error, i ...any) {
```

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/machineid/machineidv1/experiment"
+	"github.com/gravitational/teleport/lib/auth/native"
+)
+
+// TestWorkloadIdentityService_SignX509SVIDs is an integration test that uses a
+// real gRPC client/server.
+func TestWorkloadIdentityService_SignX509SVIDs(t *testing.T) {
+	experiment.SetEnabled(true)
+	t.Parallel()
```
Comment (jakule) [inline lib/auth/machineid/machineidv1/workload_identity_service_test.go:42]: If the test sets a global variable, then it cannot run in parallel. The logic will break the moment when someone adds another test that changes this flag.

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	machineidv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/machineid/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/machineid/machineidv1/experiment"
+	"github.com/gravitational/teleport/lib/auth/native"
+)
+
+// TestWorkloadIdentityService_SignX509SVIDs is an integration test that uses a
+// real gRPC client/server.
+func TestWorkloadIdentityService_SignX509SVIDs(t *testing.T) {
```
Comment (jakule) [inline lib/auth/machineid/machineidv1/workload_identity_service_test.go:40]: For all the code, do we have only one test? I don't think it covers all added logic.

File: lib/auth/machineid/machineidv1/experiment/experiment.go (hunk 1 above)
Comment (strideynet) [reply lib/auth/machineid/machineidv1/experiment/experiment.go]: 7f8e6ba1cf

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go (hunk 2 above)
Comment (strideynet) [reply lib/auth/machineid/machineidv1/workload_identity_service_test.go]: 5546c1f03d

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go (hunk 4 above)
Comment (strideynet) [reply lib/auth/machineid/machineidv1/workload_identity_service_test.go:40]: There's are more unit-tests tests introduced in https://github.com/gravitational/teleport/pull/38174 / https://github.com/gravitational/teleport/pull/38172 - however - for this PR it made the most sense to me to have a higher level integration test which assures the behaviour against the specification. This test will be far more resilient to changes and isn't reliant on the implementation details of this code.

File: lib/auth/machineid/machineidv1/workload_identity_service_test.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	experimentBefore := experiment.Enabled()
+	experiment.SetEnabled(true)
+	defer experiment.SetEnabled(experimentBefore)
```
Comment (jakule) [inline lib/auth/machineid/machineidv1/workload_identity_service_test.go:43]: nit:
```suggestion
	t.Cleanup(func() {experiment.SetEnabled(experimentBefore)})
```

**Review by jakule — APPROVED** [review]

**Review by timothyb89 — APPROVED** [review]
Apologies for the delay, just have a quick nit.

File: lib/auth/clt.go [hunk 6]
```diff
@@ -966,6 +966,12 @@ type ClientI interface {
 	// "not implemented" errors (as per the default gRPC behavior).
 	ExternalAuditStorageClient() *externalauditstorage.Client
 
+	// WorkloadIdentityServiceClient returns a workload identity service client.
+	// Clients connecting to  older Teleport versions, still get a bot service client
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/clt.go:1022]: Good catch

**Review by ryanclark — APPROVED** [review]
bot

**Pull request merged**